README.md

```diff
@@ -648,6 +648,15 @@ Another package may select this version by specifying:
 
 Also, `/// <reference types=".." />` will not work with path mapping, so dependencies must use `import`.
 
+#### How do breaking type changes work if type declaration packages closely track the library package's version?
+
+Contributors should make an effort to not ship breaking type changes that break with [semantic versioning](https://semver.org/) guidelines. However, changes to TypeScript can lead to _rare_ exceptions. Here are some examples:
```

Comment:
I feel like this puts way too much weight on "changes to TypeScript" being the motivator for breaks in a DT package. My experience has been that it's much more likely for a types maintainer to want to improve their types or fix them when they're wrong, in such a way that it must be a break. And that has nothing to do with TypeScript changing.

The two examples below are certainly rare, but that can be true regardless of the fact that there are other intentional breaks that I feel are much more common.

Reply:
Makes sense! I defer to your perspective as to what is most common, as a recent contributor :)

How about 2 bullets, one about contributors improving types in a way that they must break + vague one about changes to typescript ecosystem:


* The types maintainer improves their types or fixes them when they're wrong, in such a way that the change must be breaking.
* The the types are changed to evolve with the TypeScript ecosystem in a way that must be breaking (e.g. new module system support, TypeScript version EOL, etc.).

Reply:
I think the question is "who are these docs for?"

- Are these for contributors, where we want to say "try to not break the types unless you really have to?"
- Are they for downstream users who want to know why breaking changes happen (counter to semver)?

I think the two bullets you suggested are fine, but I think mainly it's the intro paragraph that would need tweaking. I'm not so sure that we can really point to semver that hard; the reality is that the major/minor version on DT must match the upstream package being typed. It's definitely _better_ if breaking changes only happen in major revisions, but they necessarily may not.

Reply:
They're mainly for consumers (and as I write that, I understand now why my intro paragraph needs tweaking):

- Package authors: adding a type declaration package dependency should be done with these caveats in mind. For example, using using a `^` package version constraint for a type declaration dependency means something a little different than doing so for a library package dependency. The former provides looser guarantees than the latter.
- TypeScript users: awareness of another way in which `@types/*` packages are the wild west a little bit (i.e. already know that types may be incomplete, behind the implementation package, etc.)

How about this then:

DefinitelyTyped consumers should be aware of the caveats that come with type declaration packages tracking the versions of their associated library packages. While contributors make an effort to not ship breaking type updates, changes to TypeScript can lead to _rare_ exceptions. Here are some examples:

`<originally proposed bullets>`

Reply:
Meh I don't like that either. I think it should just say the thing:

> DefinitelyTyped consumers should be aware that breaking changes can occasionally slip into minor and patch updates to type declaration packages, since they track the versions of their associated library packages. While contributors make an effort to not ship breaking type updates, changes to TypeScript can lead to _rare_ exceptions. Here are some examples:
> 
> `<originally proposed bullets>`

Reply:
If the target is consumers of `@types` packages, then I'm not _super_ sold on it being in this README in that only the top section about "what are types" and "what is the support window" is actually intended to be read by people who aren't contributing to the repo. But, I'm not totally sure. I guess it comes down to what concrete problem you're trying to solve by documenting it.

Reply:
That's totally fair. I'm not expecting all consumers to read it, but I think having it to link/point to would be helpful when people file tickets about types breaking. My point in the PR description is that I think you can put this information together by carefully reading the readme, but if I can just link people if/when types break in this specific way, having it spelled out on the front page would be useful.

The motivating factor for me is that I shipped my first of many proposed changes to types packages to improve `module: "NodeNext"` support for packages my team uses, and as luck would have it, the first one broke types for 2 packages (both on the popular side) (no need to postmortem this here — just providing the background). People were understandably filed and issue and were like "what gives", but one suggestion was to rollback and do it again as a major version update, which made me realize they might not understand how versioning works here (and the consequences of that versioning choice as it relates to breakages).

While I did explain it, I don't have much authority here. Arguments about authority aside, it's certainly easier if I can just link to the thing that explains it on the readme. If I can do that, people can read it and be like "oh right this can happen" and the source of truth is the DefinitelyTyped front page itself.

Reply:
Having something to link to makes sense. To that end, I feel like we can more concisely say something similar to what I wrote on discord, in that we state roughly the facts that:

- `@types` packages always type packages of the same version, so `@types/foo@5.4` are for `foo@5.4`.
- As a consequence, all changes, breaking or not, are published as patch revisions, unless paired with a major/minor bump to change the package version being targeted (coincidentally or not).
- Types authors are encouraged to not break the types, and to potentially put breaking changes into major bumps (which can only happen when upstream also major bumps).
- At the end of the day, it's up to the type package owners to decide what happens to the package.

Maybe literally just those sentences combined into a paragraph? (Please bikeshed)

Reply:
I think in the rare case you need to point angry types consumers to something, it’s just as effective to point them to something whose primary audience is contributors. As a DT maintainer, I can attest that contributors are also sometimes confused by this. Most understand that the versioning tracks the upstream package, but have questions about when it’s ok to ship breaks. The real answer is that the policy is to be considerate and use good judgment, and maintainers will help make those kinds of decisions. We consider the popularity of the package, the upsides of the proposed breaking change, the effort that will be required for users to fix their code, and whether the change could reasonably be delayed until it can be synced with a major bump of the upstream library. I would be in favor of having something to that effect in writing somewhere.

Reply:
I'm happy with the "written for contributors, but understandable for consumers" direction.

How about (to replace the whole body of this section under the existing proposed header):

> `@types` packages always type packages of the same version, so `@types/foo@5.4.x` are for `foo@5.4.x`.
As a consequence, all changes, breaking or not, are published as patch revisions, unless paired with a major/minor bump to change the package version being targeted (coincidentally or not).
>
> When it comes to breaking changes, DT maintainers consider the popularity of the package, the upsides of the proposed breaking change, the effort that will be required for users to fix their code, and whether the change could reasonably be delayed until it can be synced with a major bump of the upstream library.

I feel like this combines the facts with the implicit policy.

Reply:
Looks good to me!